src/transformations.rs

```diff
@@ -169,18 +176,38 @@ pub mod share_normal_form {
         }
     }
 
+    /// Type of let-binding to introduce during the share normal form pass.
+    enum BindingType {
+        Normal,
+        Revertible,
+    }
+
+    impl Default for BindingType {
+        fn default() -> Self {
+            BindingType::Normal
+        }
+    }
+
     /// Bind a list of pairs `(identifier, term)` in a term.
     ///
     /// Given the term `body` and bindings of identifiers to terms represented as a list of pairs
     /// `(id_1, term_1), .., (id_n, term_n)`, return the new term `let id_n = term_n in ... let
     /// id_1 = term_1 in body`.
-    fn with_bindings(body: Term, bindings: Vec<(Ident, RichTerm)>, pos: TermPos) -> RichTerm {
+    fn with_bindings(
+        body: Term,
+        bindings: Vec<(Ident, RichTerm)>,
+        pos: TermPos,
+        btype: BindingType,
+    ) -> RichTerm {
         bindings.into_iter().fold(
             RichTerm {
                 term: Box::new(body),
                 pos: pos.into_inherited(),
             },
-            |acc, (id, t)| RichTerm::new(Term::Let(id, t, acc), pos),
+            |acc, (id, t)| match btype {
+                BindingType::Normal => RichTerm::new(Term::Let(id, t, acc), pos),
+                BindingType::Revertible => RichTerm::new(Term::LetRev(id, t, acc), pos),
```

Comment:
This is a bit meh. Does it indicate that `Let` and `LetRev` ought to be the same `Let` with a `BindingType` argument?